apps/emqx/src/emqx_tls_certfile_gc.erl

```diff
@@ -0,0 +1,335 @@
+    start_link/0,
+    start_link/1,
+    run/0,
+    force/0
+]).
+
+-behaviour(gen_server).
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2
+]).
+
+%% Testing & maintenance
+-export([
+    orphans/0,
+    orphans/1,
+    convicts/2,
+    collect_files/2
+]).
+
+-define(GC_INVERVAL, 5 * 60 * 1000).
+
+-define(HAS_OWNER_READ(Mode), ((Mode band 8#00400) > 0)).
+-define(HAS_OWNER_WRITE(Mode), ((Mode band 8#00200) > 0)).
+
+-type filename() :: string().
+-type fileinfo() :: #file_info{}.
+
+-type st() :: #{
+    orphans => #{filename() => fileinfo()},
+    gc_interval => pos_integer(),
+    next_gc_timer => reference()
+}.
+
+-type event() ::
+    {collect, filename(), ok | {error, file:posix()}}.
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+-spec start_link() ->
+    {ok, pid()}.
+start_link() ->
+    start_link(?GC_INVERVAL).
+
+-spec start_link(_Interval :: pos_integer()) ->
+    {ok, pid()}.
+start_link(Interval) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, Interval, []).
+
+-spec run() ->
+    {ok, _Events :: [event()]}.
+run() ->
+    gen_server:call(?MODULE, collect).
+
+-spec force() ->
+    {ok, _Events :: [event()]}.
+force() ->
+    % NOTE
+    % Simulate a complete GC cycle by running it twice. Mostly useful in tests.
+    {ok, Events1} = run(),
+    {ok, Events2} = run(),
+    {ok, Events1 ++ Events2}.
+
+%%--------------------------------------------------------------------
+%% Supervisor callbacks
+%%--------------------------------------------------------------------
+
+-spec init(_) ->
+    {ok, st()}.
+init(Interval) ->
+    {ok, start_timer(#{gc_interval => Interval})}.
+
+-spec handle_call(collect | _Call, gen_server:from(), st()) ->
+    {reply, {ok, [event()]}, st()} | {noreply, st()}.
+handle_call(collect, From, St) ->
+    {ok, Events, StNext} = ?tp_span(
+        tls_certfile_gc_manual,
+        #{caller => From},
+        collect(St, #{evhandler => {fun emqx_utils:cons/2, []}})
+    ),
+    {reply, {ok, Events}, restart_timer(StNext)};
+handle_call(Call, From, St) ->
+    ?SLOG(error, #{msg => "unexpected_call", call => Call, from => From}),
+    {noreply, St}.
+
+-spec handle_cast(_Cast, st()) ->
+    {noreply, st()}.
+handle_cast(Cast, State) ->
+    ?SLOG(error, #{msg => "unexpected_cast", cast => Cast}),
+    {noreply, State}.
+
+-spec handle_info({timeout, reference(), collect}, st()) ->
+    {noreply, st()}.
+handle_info({timeout, TRef, collect}, St = #{next_gc_timer := TRef}) ->
+    {ok, _, StNext} = ?tp_span(
+        tls_certfile_gc_periodic,
+        #{},
+        collect(St, #{evhandler => {fun log_event/2, []}})
+    ),
+    {noreply, start_timer(StNext)}.
+
+start_timer(St = #{gc_interval := Interval}) ->
+    TRef = erlang:start_timer(Interval, self(), collect),
+    St#{next_gc_timer => TRef}.
+
+restart_timer(St = #{next_gc_timer := TRef}) ->
```

Comment:
ensure the old is cancelled before start a new one.
because the `col.lect` message can be sent by anyone.